### PR TITLE
[MV-478] Port Image Transformation Ops

### PR DESCRIPTION
**Pull request by o2buzzle**
This PR is opened to track the progress on porting Image Transformation group of functions of Torch Vision from MoDNN to MIOpen.

Ported Kernels:
- `ImageAdjustBrightness` (`ImageAdjustBrightnessFwd` and `ImageAdjustBrightnessFwdContiguous`)
- `ImageAdjustHue` (Including `ImageAdjustHue` and `ImageAdjustHuecontiguous`)
- `ImageAdjustSaturation` (`RGBToGrayscale`, `RGBToGrayscaleContiguous`, `Blend` and `BlendContiguous`)
- `ImageNormalize` (`ImageNormalizeFwd` and `ImageNormalizeFwdContiguous`)

Detailed performance results are  [here](https://morehio-my.sharepoint.com/:x:/r/personal/kyeonghwan_ryu_moreh_io/_layouts/15/Doc.aspx?sourcedoc=%7BD333B873-9E92-45CA-9733-7CE555199B4E%7D&file=image_transform_ops_rocm_miopen.xlsx&action=default&mobileredirect=true&wdsle=0)

Average performance:

### ImageAdjustBrightness

||contiguous|noncontiguous|
|-|----------|-------------|
|float32|	3.71689190878005|	0.723149363334591|
|float16|	3.15438010349232|	0.634670903751661|
|bfloat16|	3.32543554821554|	0.671919408850448|

### ImageAdjustHue	

||contiguous	|noncontiguous|
|-|-----------------------------|----------------------------|
|float32	|190.43048949976	|28.1578786423446|
|float16	|197.85261022511	|32.9095894372178|
|bfloat16	|214.060946381119	|32.1781814123268|

### ImageAdjustSaturation	
||contiguous	|noncontiguous|
|-|-------------|-------------|
|float32	|5.20695012592746	|0.61691319772421
|float16	|2.43261057482474	|0.704746182620516
|bfloat16	|2.3418133755682	|0.70408398450941

### ImageNormalize	

||contiguous|	noncontiguous|
|-|-------------|-------------|
|float32	|1.97416547217027	|0.538143875753033
|float16	|18.1411108179892	|0.596732511267096
|bfloat16	|1.88869907895238	|0.613813614449204

### CODE REVIEW

File: driver/CMakeLists.txt
Comment (hieule88) [inline driver/CMakeLists.txt:1]: Please move it up one line

File: driver/image_adjust_brightness_driver.hpp [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+
+    float brightness_factor;
+};
+
+template <typename Tgpu, typename Tref>
+int ImageAdjustBrightnessDriver<Tgpu, Tref>::AddCmdLineArgs()
+{
+    inflags.AddInputFlag("forw", 'F', "1", "Run only the forward pass", "int");
+
+    inflags.AddTensorFlag("input", 'I', "1x3x96x96", "Input Tensor Size");
```
Comment (hieule88) [inline driver/image_adjust_brightness_driver.hpp:124]: Please add an example to the description of this flag for convenient usage

File: driver/image_adjust_brightness_driver.hpp [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+}
+
+template <typename Tgpu, typename Tref>
+int ImageAdjustBrightnessDriver<Tgpu, Tref>::GetandSetData()
+{
+    TensorParameters input_vec = inflags.GetValueTensor("input");
+    assert(input_vec.lengths.size() == 4 || input_vec.lengths.size() == 3);
+    if(input_vec.lengths.size() == 3)
+    {
+        // n=1
```
Comment (hieule88) [inline driver/image_adjust_brightness_driver.hpp]: Please remove comment

File: driver/image_adjust_brightness_driver.hpp [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+                                    brightness_factor);
+    return miopenStatusSuccess;
+}
+
+template <typename Tgpu, typename Tref>
+int ImageAdjustBrightnessDriver<Tgpu, Tref>::VerifyForward()
+{
+    RunForwardCPU();
+
+    auto threashold = sizeof(Tgpu) == 4 ? 1e-6 : 5e-2;
```
Comment (hieule88) [inline driver/image_adjust_brightness_driver.hpp]: Is there any reason you use these constants as threasholds? I often use 
```
std::numeric_limits<Tgpu>::epsilon()
```

File: driver/image_adjust_hue_driver.hpp [hunk 4]
```diff
@@ -0,0 +1,389 @@
...
+        *g = p;
+        *b = v;
+        break;
+    case 5:
+        *r = v;
+        *g = p;
+        *b = q;
+        break;
+    default:
+        // Panic Get: How Did We Get Here?
```
Comment (hieule88) [inline driver/image_adjust_hue_driver.hpp]: Please remove comment

File: driver/image_adjust_hue_driver.hpp [hunk 5]
```diff
@@ -0,0 +1,389 @@
...
+
+    float hue;
+};
+
+template <typename Tgpu, typename Tref>
+int ImageAdjustHueDriver<Tgpu, Tref>::AddCmdLineArgs()
+{
+    inflags.AddInputFlag("forw", 'F', "1", "Run only the forward pass", "int");
+
+    inflags.AddTensorFlag("input", 'I', "1x3x96x96", "Input Tensor Size");
```
Comment (hieule88) [inline driver/image_adjust_hue_driver.hpp:216]: Please add example to description

File: driver/image_adjust_hue_driver.hpp [hunk 6]
```diff
@@ -0,0 +1,389 @@
...
+}
+
+template <typename Tgpu, typename Tref>
+int ImageAdjustHueDriver<Tgpu, Tref>::GetandSetData()
+{
+    TensorParameters input_vec = inflags.GetValueTensor("input");
+    assert(input_vec.lengths.size() == 4 || input_vec.lengths.size() == 3);
+    if(input_vec.lengths.size() == 3)
+    {
+        // n=1
```
Comment (hieule88) [inline driver/image_adjust_hue_driver.hpp]: Please remove comment

File: driver/image_adjust_hue_driver.hpp [hunk 7]
```diff
@@ -0,0 +1,389 @@
...
+        return -1;
+    }
+
+    return miopenStatusSuccess;
+}
+
+template <typename Tgpu, typename Tref>
+int ImageAdjustHueDriver<Tgpu, Tref>::RunBackwardGPU()
+{
+    // Does not exist
```
Comment (hieule88) [inline driver/image_adjust_hue_driver.hpp:341]: Please remove comment

File: driver/image_adjust_saturation_driver.hpp [hunk 8]
```diff
@@ -0,0 +1,378 @@
...
+        int n, c, h, w;
+        getNCHW(n, c, h, w, gid, dst_tv.size);
+
+        Tref r = get4DValueAt(src, src_tv, n, 0, h, w);
+        Tref g = get4DValueAt(src, src_tv, n, 1, h, w);
+        Tref b = get4DValueAt(src, src_tv, n, 2, h, w);
+
+        Tref value = 0.2989f * r + 0.587f * g + 0.114f * b;
+
+        // We expect the workspace here to always stay contiguous
```
Comment (hieule88) [inline driver/image_adjust_saturation_driver.hpp:66]: Please remove comment

File: driver/image_adjust_saturation_driver.hpp [hunk 9]
```diff
@@ -0,0 +1,378 @@
...
+                                     miopen::TensorDescriptor outputDesc,
+                                     const Tgpu* input,
+                                     Tref* output,
+                                     float saturation_factor)
+
+{
+    tensor_view_4d_t input_tv  = get_inner_expanded_4d_tv(inputDesc);
+    tensor_view_4d_t output_tv = get_inner_expanded_4d_tv(outputDesc);
+
+    // temporary view for workspace (basically a contiguous vector with same size as input_tv)
```
Comment (hieule88) [inline driver/image_adjust_saturation_driver.hpp:110]: Please remove comment

File: driver/image_adjust_saturation_driver.hpp [hunk 10]
```diff
@@ -0,0 +1,378 @@
...
+
+    float saturation_factor;
+};
+
+template <typename Tgpu, typename Tref>
+int ImageAdjustSaturationDriver<Tgpu, Tref>::AddCmdLineArgs()
+{
+    inflags.AddInputFlag("forw", 'F', "1", "Run only the forward pass", "int");
+
+    inflags.AddTensorFlag("input", 'I', "1x3x96x96", "Input Tensor Size");
```
Comment (hieule88) [inline driver/image_adjust_saturation_driver.hpp:192]: Please add example to description

File: driver/image_adjust_saturation_driver.hpp [hunk 11]
```diff
@@ -0,0 +1,378 @@
...
+}
+
+template <typename Tgpu, typename Tref>
+int ImageAdjustSaturationDriver<Tgpu, Tref>::GetandSetData()
+{
+    TensorParameters input_vec = inflags.GetValueTensor("input");
+    assert(input_vec.lengths.size() == 4 || input_vec.lengths.size() == 3);
+    if(input_vec.lengths.size() == 3)
+    {
+        // n=1
```
Comment (hieule88) [inline driver/image_adjust_saturation_driver.hpp]: Please remove comment

File: driver/image_normalize_driver.hpp [hunk 12]
```diff
@@ -0,0 +1,347 @@
...
+    std::vector<Tgpu> stdvar_host;
+
+    std::vector<Tref> out_ref;
+};
+
+template <typename Tgpu, typename Tref>
+int ImageNormalizeDriver<Tgpu, Tref>::AddCmdLineArgs()
+{
+    inflags.AddInputFlag("forw", 'F', "1", "Run only the forward pass", "int");
+    inflags.AddTensorFlag("input", 'I', "1x3x96x96", "Input Tensor Size");
```
Comment (hieule88) [inline driver/image_normalize_driver.hpp:137]: Please add example to description

File: include/miopen/miopen.h [hunk 13]
```diff
@@ -512,7 +513,7 @@ typedef enum
     miopenActivationABS      = 5, /*!< Absolute value \f$abs(x)\f$ */
     miopenActivationPOWER = 6, /*!< Scaled and shifted power \f$(\alpha + \beta * x)^{gamma}\f$ */
     miopenActivationCLIPPEDRELU =
-        7, /*!< Clipped Rectified Linear Unit \f$ min(\alpha, max(0,x)) \f$ */
```
Comment (hieule88) [inline include/miopen/miopen.h]: Please remove unrelated change

File: driver/image_normalize_driver.hpp [hunk 14]
```diff
@@ -0,0 +1,347 @@
...
+
+template <typename Tgpu, typename Tref>
+int ImageNormalizeDriver<Tgpu, Tref>::GetandSetData()
+{
+    TensorParameters input_vec = inflags.GetValueTensor("input");
+
+    assert(input_vec.lengths.size() == 4 || input_vec.lengths.size() == 3);
+    if(input_vec.lengths.size() == 3)
+    {
+        // n=1
```
Comment (hieule88) [inline driver/image_normalize_driver.hpp]: Please remove comment

File: driver/image_normalize_driver.hpp [hunk 15]
```diff
@@ -0,0 +1,347 @@
...
+
+    auto strides = ComputeStrides(input_vec.lengths, inflags.GetValueInt("contiguous") == 1);
+
+    SetTensorNd(inputTensorDesc, input_vec.lengths, strides, data_type);
+    SetTensorNd(outputTensorDesc, input_vec.lengths, strides, data_type);
+
+    TensorParameters mean_vec   = inflags.GetValueTensor("mean");
+    TensorParameters stdvar_vec = inflags.GetValueTensor("stdvar");
+
+    // Assert that these are at least as large as the input channels
```
Comment (hieule88) [inline driver/image_normalize_driver.hpp:183]: Please remove comment

File: src/CMakeLists.txt [hunk 16]
```diff
@@ -260,6 +263,11 @@ set( MIOpen_Source
     solver/gemm_bwd.cpp
     solver/gemm_wrw.cpp
     solver/groupnorm/forward_groupnorm.cpp
+    solver/image_transform/adjust_brightness.cpp
+    solver/image_transform/adjust_hue.cpp
+    solver/image_transform/adjust_contrast.cpp
```
Comment (hieule88) [inline src/CMakeLists.txt]: Please move it up one line

File: src/solver/image_transform/adjust_contrast.cpp [hunk 17]
```diff
@@ -0,0 +1,165 @@
...
+    result.invoker_factory = [](const std::vector<Kernel>& kernels) {
+        return [=](const Handle& handle, const AnyInvokeParams& raw_params) {
+            decltype(auto) params =
+                raw_params.CastTo<miopen::image_transform::adjust_contrast::InvokeParams>();
+
+            auto kernel = handle.Run(kernels[0]); // RGB to Grayscale
+
+            auto xdesc = miopen::deref(params.inputTensorDesc);
+
+            // WS is a contiguous tensor with the same shape as the input
```
Comment (hieule88) [inline src/solver/image_transform/adjust_contrast.cpp]: Please remove comment

**Review by hieule88 — COMMENTED** [review]
Partial reviews

File: include/miopen/miopen.h (hunk 13 above)
Comment (o2buzzle) [reply include/miopen/miopen.h]: Removed the erroneous formatting

File: driver/image_adjust_brightness_driver.hpp (hunk 1 above)
Comment (o2buzzle) [reply driver/image_adjust_brightness_driver.hpp:124]: Since the flag is `AddTensorFlag`, they are added automatically by the framework

image

File: driver/image_adjust_brightness_driver.hpp (hunk 3 above)
Comment (o2buzzle) [reply driver/image_adjust_brightness_driver.hpp]: Check the other drivers for details. It's seems acceptable to use custom values here if desired.

File: driver/image_adjust_hue_driver.hpp (hunk 5 above)
Comment (o2buzzle) [reply driver/image_adjust_hue_driver.hpp:216]: Similar to the other drivers. It is added by the framework

File: driver/image_adjust_hue_driver.hpp (hunk 7 above)
Comment (o2buzzle) [reply driver/image_adjust_hue_driver.hpp:341]: This is probably necessary in order to note that the backward functions for these do not exist, and thus running them will simply return. The function must exist because it is an abstract function of the parent class

File: driver/image_adjust_hue_driver.hpp (hunk 6 above)
Comment (o2buzzle) [reply driver/image_adjust_hue_driver.hpp]: Changed the comment to better reflect why an insertion of `n=1` is necessary

File: driver/image_normalize_driver.hpp (hunk 14 above)
Comment (o2buzzle) [reply driver/image_normalize_driver.hpp]: Changed the comment to better reflect why an insertion of `n=1` is necessary

File: driver/image_adjust_saturation_driver.hpp (hunk 8 above)
Comment (o2buzzle) [reply driver/image_adjust_saturation_driver.hpp:66]: I think this kind of comments in the framework are allowed.

File: driver/image_adjust_saturation_driver.hpp (hunk 11 above)
Comment (o2buzzle) [reply driver/image_adjust_saturation_driver.hpp]: Changed the comment to better reflect why an insertion of `n=1` is necessary

File: driver/image_adjust_brightness_driver.hpp (hunk 3 above)
Comment (hieule88) [reply driver/image_adjust_brightness_driver.hpp]: Is it oke @et16kr

File: src/CMakeLists.txt (hunk 16 above)
Comment (o2buzzle) [reply src/CMakeLists.txt]: Reordered as requested

File: driver/image_adjust_hue_driver.hpp (hunk 4 above)
Comment (o2buzzle) [reply driver/image_adjust_hue_driver.hpp]: Changed the comment to make this part clear

File: driver/image_adjust_saturation_driver.hpp (hunk 10 above)
Comment (o2buzzle) [reply driver/image_adjust_saturation_driver.hpp:192]: Same as other drivers. This will be added by the framework itself

File: driver/image_normalize_driver.hpp (hunk 12 above)
Comment (o2buzzle) [reply driver/image_normalize_driver.hpp:137]: Same as other drivers. This will be added by the framework itself

File: driver/image_adjust_brightness_driver.hpp (hunk 3 above)
Comment (o2buzzle) [reply driver/image_adjust_brightness_driver.hpp]: As far as I understand you *can* set your own threashold (check `pool_driver`). I've changed to `numeric_limits::epsilon`

File: src/solver/image_transform/adjust_saturation.cpp [hunk 18]
```diff
@@ -0,0 +1,220 @@
...
+    size_t zlocalsize = 1;
+    size_t zgridsize  = 1;
+
+    // first, RGB to Grayscale
+    auto rgb_kernel        = KernelInfo{};
+    rgb_kernel.kernel_file = "MIOpenImageBlend.cpp";
+    // if(problem.GetInputTensorDesc().IsContiguous() &&
+    // problem.GetOutputTensorDesc().IsContiguous())
+    //     rgb_kernel.kernel_name = "RGBToGrayscaleContiguous";
+    // else
```
Comment (et16kr) [inline src/solver/image_transform/adjust_saturation.cpp]: Please clean up the comments.

File: src/solver/image_transform/adjust_hue.cpp [hunk 19]
```diff
@@ -0,0 +1,140 @@
...
+            auto xdesc = miopen::deref(params.inputTensorDesc);
+            auto ydesc = miopen::deref(params.outputTensorDesc);
+
+            auto x_tv = get_inner_expanded_4d_tv(xdesc);
+            auto y_tv = get_inner_expanded_4d_tv(ydesc);
+
+            size_t N        = xdesc.GetElementSize() / 3;
+            size_t c_stride = xdesc.GetLengths()[2] * xdesc.GetLengths()[3];
+
+            if(kernel.name == "ImageAdjustHueContiguous")
```
Comment (et16kr) [inline src/solver/image_transform/adjust_hue.cpp:120]: Since contiguous is distinguished in NetworkConfig(), it might be better to create a separate invoker_factory and remove the if statement.

File: test/gtest/image_adjust_brightness.cpp [hunk 20]
```diff
@@ -0,0 +1,68 @@
...
+{
+};
+
+} // namespace image_adjust_brightness
+
+using namespace image_adjust_brightness;
+
+TEST_P(ImageAdjustBrightnessTestFloat, ImageAdjustBrightnessTestFw)
+{
+    if(miopen::IsEnabled(ENV(MIOPEN_TEST_ALL)) && (GetFloatArg() == "--float"))
```
Comment (et16kr) [inline test/gtest/image_adjust_brightness.cpp]: Please refer to the link and add the following. (https://github.com/moreh-dev/MIOpen/pull/27#discussion_r1634984024)
```miopen::IsUnset(ENV(MIOPEN_TEST_ALL))```

File: src/image_transform.cpp [hunk 21]
```diff
@@ -0,0 +1,245 @@
...
+    const auto algo = AlgorithmName{"ImageNormalize"};
+    const auto solver =
+        solver::SolverContainer<solver::image_transform::normalize::ImageNormalize>{};
+
+    solver.ExecutePrimitive(handle, problem, algo, invoke_params);
+
+    return miopenStatusSuccess;
+}
+
+miopenStatus_t miopenImageAdjustContrast(Handle& handle,
```
Comment (et16kr) [inline src/image_transform.cpp]: Is ImageAdjustContrast being used?
If not, please remove it along with the solver.
If it is being used, please add it to miopen.h and other relevant files.

File: src/include/miopen/image_transform/adjust_brightness/problem_description.hpp [hunk 22]
```diff
@@ -0,0 +1,94 @@
...
+        if(inputTensorDesc.GetLengths().size() == 4)
+            return true;
+
+        return false;
+    }
+
+    bool IsImprovementOverROCm() const
+    {
+        if(inputTensorDesc.IsContiguous() && outputTensorDesc.IsContiguous())
+            return true;
```
Comment (et16kr) [inline src/include/miopen/image_transform/adjust_brightness/problem_description.hpp:82]: If it does not work with non-contiguous, please remove the unused non-contiguous kernels.
If they need to be kept for testing or other purposes, please store them in a separate branch.
When submitting the PR to AMD, unused kernels should not be included.

File: src/include/miopen/image_transform/adjust_hue/problem_description.hpp [hunk 23]
```diff
@@ -0,0 +1,105 @@
...
+    bool IsHueInRange() const { return hue >= -0.5f && hue <= 0.5f; }
+
+    bool IsInputSizesValid()
+    {
+        // We can only really support 4d tensors (ala. NCHW).
+        if(inputTensorDesc.GetLengths().size() != 4)
+            return false;
+
+        // This op only support colored images
+        if(inputTensorDesc.GetLengths()[1] != 3)
```
Comment (et16kr) [inline src/include/miopen/image_transform/adjust_hue/problem_description.hpp:87]: This condition is also needed for image_adjust_saturation.
There are a few other things that need to be checked.
```
// image_adjust_saturation
MODNN_CHECK_ARGUMENT(saturation_factor >= 0.0,
                     "op=ImageAdjustSaturation, Saturation factor must be "
                     "greater than or equal to 0");
// image normalize
MODNN_CHECK_ARGUMENT(mean_tensor.is_contiguous() && std_tensor.is_contiguous(),
                     "op=ImageNormalize, mean and std tensor must be contiguous for "
                     "ImageNormalize");
```
Please refer to modnn and add them.

File: src/image_transform.cpp [hunk 24]
```diff
@@ -0,0 +1,245 @@
...
+#include "miopen/image_transform/normalize/problem_description.hpp"
+#include "miopen/image_transform/solvers.hpp"
+#include "miopen/miopen.h"
+#include "miopen/names.hpp"
+#include "miopen/tensor.hpp"
+#include <cstddef>
+
+namespace miopen {
+
+miopenStatus_t miopenImageAdjustHue(Handle& handle,
```
Comment (et16kr) [inline src/image_transform.cpp]: Please modify the function names in this file as follows:
miopen::miopenImageAdjustHue() to miopen::ImageAdjustHue()

File: src/image_transform.cpp (hunk 21 above)
Comment (o2buzzle) [reply src/image_transform.cpp]: `ImageAdjustContrast` is meant to be added later as a branch of this PR. The issue is that it basically is `ImageAdjustSaturation`, but with an extra calculation of the mean value of the image for blending, which appears to not be available in MIOpen yet. I've asked Kyeonghwan if I can split this PR into two as the scope of ImageAdjustContrast means I would have to port `common::Mean` from the MoDNN as well.

File: src/include/miopen/image_transform/adjust_hue/problem_description.hpp (hunk 23 above)
Comment (o2buzzle) [reply src/include/miopen/image_transform/adjust_hue/problem_description.hpp:87]: I've added the necessary checks

File: src/include/miopen/image_transform/adjust_brightness/problem_description.hpp (hunk 22 above)
Comment (o2buzzle) [reply src/include/miopen/image_transform/adjust_brightness/problem_description.hpp:82]: I am currently using them on a separate branch to cross check accuracy with the contiguous functions to ensure my changes doesn't introduce new issues. I will remove them after the other reviews are resolved

File: test/gtest/image_adjust_brightness.cpp (hunk 20 above)
Comment (o2buzzle) [reply test/gtest/image_adjust_brightness.cpp]: I've added the envvar check

File: src/image_transform.cpp (hunk 24 above)
Comment (o2buzzle) [reply src/image_transform.cpp]: I've renamed the internal APIs

**Comment by o2buzzle** [comment]
@et16kr performance on ImageAdjustContrast after porting is looking worse than I initially expect, given the MoDNN results (ported ops avg. ~1.1x for contiguous float32, worse for non-contiguous cases). I'll do some extra tests to see if there are any cases that could potentially give acceptable performance, but if not, I'll cleanup the PR and get it ready by tomorrow.

File: src/image_transform.cpp (hunk 21 above)
Comment (o2buzzle) [reply src/image_transform.cpp]: I've removed every mentions of it from the code, as it is not performant enough for us to port

File: src/solver/image_transform/adjust_hue.cpp (hunk 19 above)
Comment (o2buzzle) [reply src/solver/image_transform/adjust_hue.cpp:120]: I mean either way it is going to make two `invoker_factory` object anyway, so I doubt it is going to have much impact. If necessary, I can perform the necessary changes, but since these kernels have really similar launch parameters, I think this approach would be cleaner and avoids code duplication.

Also, this is exactly how it is implemented in MoDNN

File: src/solver/image_transform/adjust_saturation.cpp (hunk 18 above)
Comment (o2buzzle) [reply src/solver/image_transform/adjust_saturation.cpp]: I've removed them

File: src/include/miopen/image_transform/adjust_brightness/problem_description.hpp (hunk 22 above)
Comment (o2buzzle) [reply src/include/miopen/image_transform/adjust_brightness/problem_description.hpp:82]: I've removed the unused kernels

File: src/solver/image_transform/adjust_hue.cpp (hunk 19 above)
Comment (et16kr) [reply src/solver/image_transform/adjust_hue.cpp:120]: OK!

File: src/include/miopen/image_transform/adjust_hue/problem_description.hpp (hunk 23 above)
Comment (et16kr) [reply src/include/miopen/image_transform/adjust_hue/problem_description.hpp:87]: I made an unclear request earlier. I apologize.

```if(inputTensorDesc.GetLengths()[1] != 3)``` 
This condition is also needed for image_adjust_saturation.
[modnn/core/op/image_adjust_saturation.cpp:174](https://github.com/moreh-dev/backend/blob/8a6d54f3215495fde0cd13497d2d1578bc115bb4/modnn/core/op/image_adjust_saturation.cpp#L174)
```
  MODNN_CHECK_ARGUMENT(input.size(1) == 3,
                       "op=ImageAdjustSaturation, ImageAdjustSaturation only "
                       "supports RGB images");
```

File: src/include/miopen/image_transform/adjust_saturation/invoke_params.hpp [hunk 25]
```diff
@@ -0,0 +1,55 @@
...
+    ConstData_t input_buf = nullptr;
+    Data_t workspace_buf  = nullptr;
+    Data_t output_buf     = nullptr;
+
+    float saturation_factor = 0.0f;
+
+    size_t workspace_size = 0;
+
+    size_t GetWorkspaceSize() const { return workspace_size; }
+    Data_t GetWorkspace() const { return nullptr; }
```
Comment (et16kr) [inline src/include/miopen/image_transform/adjust_saturation/invoke_params.hpp]: If workspace is used, maybe the general usage seems to be like this. 
```Data_t GetWorkspace() const { return workspace_buf; }```

Comment (o2buzzle) [reply src/include/miopen/image_transform/adjust_saturation/invoke_params.hpp]: I've changed it to return the correct buffer

File: driver/driver.hpp [hunk 26]
```diff
@@ -151,7 +151,9 @@ inline void PadBufferSize(size_t& sz, int datatype_sz)
            "pool[fp16], lrn[fp16], "
            "activ[fp16], softmax[fp16], bnorm[fp16], rnn[fp16], gemm[fp16], ctc, dropout[fp16], "
            "tensorop[fp16], reduce[fp16|fp64], layernorm[bfp16|fp16], sum[bfp16|fp16], "
-           "argmax[bfp16|fp16], groupnorm[bfp16|fp16], cat[bfp16|fp16]\n");
+           "argmax[bfp16|fp16], groupnorm[bfp16|fp16], cat[bfp16|fp16], "
+           "image_adjust_hue[_fp16|_bfp16], image_adjust_saturation[_fp16|_bfp16], "
+           "image_normalize[_fp16|_bfp16] \n");
```
Comment (HoangNguyenMoreh) [inline driver/driver.hpp]: I think you should follow the naming convention and remove the "_".
There're also existing some long naming op in AMD code base aswell https://github.com/ROCm/MIOpen/blob/develop/driver/driver.hpp

File: driver/image_adjust_brightness_driver.hpp [hunk 27]
```diff
@@ -0,0 +1,301 @@
...
+    }
+
+    return miopenStatusSuccess;
+}
+
+template <typename Tgpu, typename Tref>
+int ImageAdjustBrightnessDriver<Tgpu, Tref>::RunBackwardGPU()
+{
+    // Does not exist
+    return miopenStatusSuccess;
```
Comment (HoangNguyenMoreh) [inline driver/image_adjust_brightness_driver.hpp:284]: Did you try `miopenStatusNotImplemented` or `miopenStatusUnsupportedOp` ?

File: driver/image_adjust_driver_common.hpp [hunk 28]
```diff
@@ -0,0 +1,52 @@
...
+    if(!contiguous)
+        std::swap(strides.front(), strides.back());
+    return strides;
+}
+
+template <typename T>
+T clamp(T val, T min, T max)
+{
+    val = val < min ? min : val;
+    val = val > max ? max : val;
```
Comment (HoangNguyenMoreh) [inline driver/image_adjust_driver_common.hpp]: did you try `std::clamp(val, min, max)` ?

Comment (o2buzzle) [reply driver/image_adjust_driver_common.hpp]: Done, thank for that suggestion

File: driver/image_adjust_brightness_driver.hpp (hunk 27 above)
Comment (o2buzzle) [reply driver/image_adjust_brightness_driver.hpp:284]: It has to be that way (check the Sum driver for a similar case where backwards doesn't exist)

File: driver/driver.hpp (hunk 26 above)
Comment (o2buzzle) [reply driver/driver.hpp]: I've removed the underscores

**Review by HoangNguyenMoreh — APPROVED** [review]

**Comment by HoangNguyenMoreh** [comment]
Looks good to me! But please don't merge the PR by yourself.

**Comment by et16kr** [comment]
prease check it. https://github.com/moreh-dev/MIOpen/pull/26#discussion_r1654302502

File: src/include/miopen/image_transform/adjust_hue/problem_description.hpp (hunk 23 above)
Comment (o2buzzle) [reply src/include/miopen/image_transform/adjust_hue/problem_description.hpp:87]: I've added the check to `adjust_saturation`

**Comment by et16kr** [comment]
I have completed my review of this PR.
If @hieule88 agrees, please close this PR.

**Comment by hieule88** [comment]
LGTM! Sorry for late reply.
